### PR TITLE
Code for ALO(Antlion Optimizer)

### PR DESCRIPTION
**Pull request by rishavpramanik**
ALO implementation, tested on 6 standard UCI Datasets. Results are listed below
Comparison Paper Link: https://doi.org/10.1016/j.neucom.2016.03.101
Fitness is tuned for the datasets
Clean-1: 100, Paper: 99
Clean-2: 100 Paper: 99.7
Zoo: 100 Paper: 99.1
Wine: 100 Paper: 100
Semeion: 94.5 Paper: 99
Sonar: 93.65 Paper 97

### CODE REVIEW

File: Py_FS/wrapper/nature_inspired/ALO.py [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+import copy
+import numpy as np
+import time
+import matplotlib.pyplot as plt
+from sklearn.datasets import fetch_openml
+from sklearn.model_selection import train_test_split
+from sklearn import datasets
+from sklearn.preprocessing import LabelEncoder
+from _utilities_test import Solution, Data, compute_accuracy, compute_fitness, initialize, sort_agents, display, call_counter
+from wrapper.nature_inspired._transfer_functions import get_trans_function
```
Comment (consigliere19) [inline Py_FS/wrapper/nature_inspired/ALO.py]: Is this import required? If it is, then change the import statement to reflect the correct directory structure.

File: Py_FS/wrapper/nature_inspired/ALO.py [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+
+class ALO():
+    def __init__(self,
```
Comment (consigliere19) [inline Py_FS/wrapper/nature_inspired/ALO.py:20]: Is ALO not inheriting the Algorithm class?

File: Py_FS/wrapper/nature_inspired/ALO.py [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+import time
+import matplotlib.pyplot as plt
+from sklearn.datasets import fetch_openml
+from sklearn.model_selection import train_test_split
+from sklearn import datasets
+from sklearn.preprocessing import LabelEncoder
+from _utilities_test import Solution, Data, compute_accuracy, compute_fitness, initialize, sort_agents, display, call_counter
+from wrapper.nature_inspired._transfer_functions import get_trans_function
+
+class ALO():
```
Comment (consigliere19) [inline Py_FS/wrapper/nature_inspired/ALO.py]: Is ALO not inheriting the Algorithm class?

**Comment by consigliere19** [comment]
**Minor Comments:**
Remove the redundant lines which have been commented.
Use blank lines between methods, follow proper formatting.

File: Py_FS/wrapper/nature_inspired/ALO.py (hunk 1 above)
Comment (rishavpramanik) [reply Py_FS/wrapper/nature_inspired/ALO.py:8]: Done!

Comment (rishavpramanik) [reply Py_FS/wrapper/nature_inspired/ALO.py:8]: Done!

Comment (rishavpramanik) [reply Py_FS/wrapper/nature_inspired/ALO.py]: Done!

File: Py_FS/wrapper/nature_inspired/ALO.py (hunk 3 above)
Comment (rishavpramanik) [reply Py_FS/wrapper/nature_inspired/ALO.py]: Done with Changes

**Comment by rishavpramanik** [comment]
Done with changes

File: Py_FS/wrapper/nature_inspired/ALO.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                        train_data=train_data,
+                        train_label=train_label,
+                        save_conv_graph=save_conv_graph,
+                        seed=seed )
+
+        self.algo_name = 'ALO'
+        self.agent_name = 'Antlion'
+        self.trans_function = trans_func_shape
+        self.algo_params = {}
+        self.popul = initialize(self.num_agents, self.train_data.shape[-1]) #initialise the ants
```
Comment (consigliere19) [inline Py_FS/wrapper/nature_inspired/ALO.py]: What is self.popul?

Comment (rishavpramanik) [reply Py_FS/wrapper/nature_inspired/ALO.py]: It is the ants, excuse I'm very bad at taking variables

Comment (trin07) [reply Py_FS/wrapper/nature_inspired/ALO.py:36]: @rishavpramanik contradiction while calling combine function in line 127.

Comment (trin07) [reply Py_FS/wrapper/nature_inspired/ALO.py]: The variables are a bit confusing. If the name for ants is population in the original paper, then instead change self.popul to self.ants

Comment (rishavpramanik) [reply Py_FS/wrapper/nature_inspired/ALO.py]: Updated

**Comment by trin07** [comment]
See in WOA_test.py, the transfer function too has been taken as input from the user. Include this option in user_input function and as algo_params.

**Comment by rishavpramanik** [comment]
> See in WOA_test.py, the transfer function too has been taken as input from the user. Include this option in user_input function and as algo_params.

Updated along with directory issues

File: Py_FS/wrapper/nature_inspired/ALO.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+class ALO(Algorithm):
+    def __init__(self,
+                 num_agents,
+                 max_iter,
+                 train_data,
+                 train_label,
+                 test_data=None,
+                 test_label=None,
+                 val_size=20,
```
Comment (trin07) [inline Py_FS/wrapper/nature_inspired/ALO.py]: What is the use of the val_size variable? Didn't find it anywhere else.

Comment (rishavpramanik) [reply Py_FS/wrapper/nature_inspired/ALO.py]: Good point, removed!

**Comment by rishavpramanik** [comment]
Hopefully resolved the issues raised

File: Py_FS/wrapper/nature_inspired/ALO.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+                    self.population[i,j] = 0
+        self.cur_iter+=1
+        compute_fitness(self.weight_acc)
+        
+        
+        
+
+if __name__ == '__main__':
+    data=datasets.load_digits()
+    alo = ALO(num_agents=50,
```
Comment (trin07) [inline Py_FS/wrapper/nature_inspired/ALO.py]: Change the number of agents to 20. This is because all other algos have 20.

Comment (rishavpramanik) [reply Py_FS/wrapper/nature_inspired/ALO.py]: Done!

**Pull request closed without merging**